Std/Data/HashMap/Lemmas.lean

```diff
@@ -0,0 +1,346 @@
+/-
+Copyright (c) 2022 Wojciech Nawrocki. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Wojciech Nawrocki
+-/
+import Std.Data.HashMap.Basic
+import Std.Data.HashMap.WF
+import Std.Data.List.Lemmas
+import Std.Data.List.Perm
+import Std.Data.Array.Lemmas
+import Std.Tactic.ShowTerm
+
+namespace Std.HashMap
+variable [BEq α] [Hashable α] [LawfulHashable α] [PartialEquivBEq α]
+
+namespace Imp
+open List
+
+-- NOTE(WN): These would ideally be solved by a congruence-closure-for-PERs tactic
+-- See https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Rewriting.20congruent.20relations
+-- Same for proofs about List.Perm
+private theorem beq_nonsense_1 {a b c : α} : a != b → a == c → b != c :=
+  fun h₁ h₂ => Bool.bne_iff_not_beq.mpr fun h₃ =>
+    Bool.bne_iff_not_beq.mp h₁ (PartialEquivBEq.trans h₂ (PartialEquivBEq.symm h₃))
+
+private theorem beq_nonsense_2 {a b c : α} : a == b → b == c → ¬(c != a) :=
+  fun h₁ h₂ h₃ => Bool.bne_iff_not_beq.mp (bne_symm h₃) (PartialEquivBEq.trans h₁ h₂)
+
+namespace Buckets
+
+/-- The contents of any given bucket are pairwise `bne`. -/
+theorem Pairwise_bne_bucket (bkts : Buckets α β) (H : bkts.WF) (i : Nat) (h : i < bkts.val.size)
+  : Pairwise (·.1 != ·.1) bkts.val[i].toList := by
+  have := H.distinct bkts.val[i] (Array.getElem_mem_data _ _)
+  exact Pairwise.imp Bool.bne_iff_not_beq.mpr this
+
+/-- Reformulation of `Pairwise_bne_bucket` for use with `List.foo_of_unique`. -/
+theorem Pairwise_bne_bucket' (bkts : Buckets α β) (H : bkts.WF) (i : Nat)
+    (h : i < bkts.val.size) (a : α)
+    : Pairwise (fun p q => p.1 == a → q.1 != a) bkts.val[i].toList :=
+  Pairwise.imp beq_nonsense_1 (Pairwise_bne_bucket bkts H i h)
+
+/-- It is a bit easier to reason about `foldl (append)` than `foldl (foldl)`, so we use this
+(less efficient) variant of `toList` as the mathematical model. -/
+def toListModel (bkts : Buckets α β) : List (α × β) :=
+  -- Note(WN): the implementation is  `bkts.foldl` rather than `bkts.data.foldl` because we need
+  -- to reason about array indices in some of the theorems.
+  bkts.val.foldl (init := []) (fun acc bkt => acc ++ bkt.toList)
+
+protected theorem foldl_cons_fn (l₁ l₂ : List α) :
```

Comment:
I made these local `simp`s since they seemed a bit too specific to put in `List.Lemmas`.

Reply:
They don't look like good global simp lemmas but I don't see any reason not to have them as regular theorems in `List.Lemmas`.